oauth/client_factory.go

```diff
@@ -14,156 +14,133 @@
-// WithAuthFromRequestIntoContext functions with clients having this type.
-type AuthenticatingClient client.Client
+var (
+	errWrongResponseContent = errors.New("unable to parse response from workspace API requested for namespace")
+	errWorkspaceNotFound    = errors.New("target workspace not found for namespace")
+	errClientInstanceNotSet = errors.New("client instance does not set")
+)
 
-type ClientFactoryConfig struct {
-	KubeInsecureTLS bool
-	KubeConfig      string
-	ApiServer       string
-	ApiServerCAPath string
+type K8sClientFactory interface {
+	CreateClient(ctx context.Context) (client.Client, error)
 }
 
-// CreateUserAuthClient creates a new client based on the provided configuration. We use this client for k8s requests with user's token (e.g.: SelfSubjectAccessReview).
-// Note that configuration is potentially modified during the call.
-func CreateUserAuthClient(args *ClientFactoryConfig) (AuthenticatingClient, error) {
-	// we can't use the default dynamic rest mapper, because we don't have a token that would enable us to connect
-	// to the cluster just yet. Therefore, we need to list all the resources that we are ever going to query using our
-	// client here thus making the mapper not reach out to the target cluster at all.
-	mapper := meta.NewDefaultRESTMapper([]schema.GroupVersion{})
-	mapper.Add(authz.SchemeGroupVersion.WithKind("SelfSubjectAccessReview"), meta.RESTScopeRoot)
-	mapper.Add(v1beta1.GroupVersion.WithKind("SPIAccessToken"), meta.RESTScopeNamespace)
-	mapper.Add(v1beta1.GroupVersion.WithKind("SPIAccessTokenDataUpdate"), meta.RESTScopeNamespace)
-
-	return createClient(args, mapper, true)
+type InClusterK8sClientFactory struct {
+	ClientOptions *client.Options
+	RestConfig    *rest.Config
 }
 
-// CreateInClusterClient creates a new client based on the provided configuration. We use this client for k8s requests with ServiceAccount (e.g.: reading configuration secrets).
-func CreateInClusterClient(args *ClientFactoryConfig) (client.Client, error) {
-	mapper := meta.NewDefaultRESTMapper([]schema.GroupVersion{})
-	mapper.Add(corev1.SchemeGroupVersion.WithKind("Secret"), meta.RESTScopeNamespace)
+type WorkspaceAwareK8sClientFactory struct {
+	ClientOptions    *client.Options
+	RestConfig       *rest.Config
+	ApiServer        string
+	WorkspaceApiPath string
+	HTTPClient       rest.HTTPClient
+}
 
-	return createClient(args, mapper, false)
+type UserAuthK8sClientFactory struct {
+	ClientOptions *client.Options
+	RestConfig    *rest.Config
 }
 
-func createClient(args *ClientFactoryConfig, mapper meta.RESTMapper, userAuthentication bool) (client.Client, error) {
-	kubeconfig, err := baseKubernetesConfig(args, userAuthentication)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create kubernetes incluster config: %w", err)
+func (w WorkspaceAwareK8sClientFactory) CreateClient(ctx context.Context) (client.Client, error) {
```

Comment:
I think it's time to split this into more files, even create new package for client factory

Reply:
can you please split this very long and complex function? it's doing like 10 different things

Reply:
it looks huge due to excessive errors processing, but generally just does the HTTP request and parses the response.
Not sure if a bunch of non-reusable fucntions will improve the situation much... how do you see it better ?

Reply:
I may propose externalize the HTTP stuff, but i'm not sure if it makes sense to make it more granular insede that.

Reply:
this file has interface and 4 structs implementing this interface. To me, this is already a reason to organize it somehow better than all in one file.

Reply:
ok i have it split per-type